pkg/terror/terror.go

```diff
@@ -0,0 +1,331 @@
+	if s, ok := errLevel2Str[el]; ok {
+		return s
+	}
+	return fmt.Sprintf("unknown error level: %d", el)
+}
+
+// Error implements error interface and add more useful fields
+type Error struct {
+	code     ErrCode
+	class    ErrClass
+	scope    ErrScope
+	level    ErrLevel
+	message  string
+	args     []interface{}
+	rawCause error
+	stack    errors.StackTracer
+}
+
+// New creates a new *Error instance
+func New(code ErrCode, class ErrClass, scope ErrScope, level ErrLevel, message string) *Error {
+	return &Error{
+		code:    code,
+		class:   class,
+		scope:   scope,
+		level:   level,
+		message: message,
+	}
+}
+
+// Code returns ErrCode
+func (e *Error) Code() ErrCode {
+	return e.code
+}
+
+// Class returns ErrClass
+func (e *Error) Class() ErrClass {
+	return e.class
+}
+
+// Scope returns ErrScope
+func (e *Error) Scope() ErrScope {
+	return e.scope
+}
+
+// Level returns ErrLevel
+func (e *Error) Level() ErrLevel {
+	return e.level
+}
+
+// Error implements error interface.
+func (e *Error) Error() string {
+	return fmt.Sprintf("[%d:%s:%s:%s] %s", e.code, e.class, e.scope, e.level, e.getMsg())
+}
+
+// Format accepts flags that alter the printing of some verbs
+func (e *Error) Format(s fmt.State, verb rune) {
+	switch verb {
+	case 'v':
+		if s.Flag('+') {
+			io.WriteString(s, e.Error())
+			if e.stack != nil {
+				e.stack.StackTrace().Format(s, verb)
+			}
+			return
+		}
+		fallthrough
+	case 's':
+		io.WriteString(s, e.Error())
+	case 'q':
+		fmt.Fprintf(s, "%q", e.Error())
+	}
+}
+
+// Cause implements causer.Cause defined in pingcap/errors
+// and returns the raw cause of an *Error
+func (e *Error) Cause() error {
+	return e.rawCause
+}
+
+func (e *Error) getMsg() string {
+	if len(e.args) > 0 {
+		return fmt.Sprintf(e.message, e.args...)
+	}
+	return e.message
+}
+
+// Equal checks if err equals to e.
+func (e *Error) Equal(err error) bool {
+	if error(e) == err {
+		return true
+	}
+	inErr, ok := err.(*Error)
+	return ok && e.code == inErr.code
+}
+
+// SetMessage clones an Error and resets its message
+func (e *Error) SetMessage(message string) *Error {
+	err := *e
+	err.message = message
+	err.args = append([]interface{}{}, e.args...)
+	return &err
+}
+
+// New generates a new *Error with the same class and code, and replace message with new message
+func (e *Error) New(message string) error {
+	return e.levelGeneratef(1, message)
+}
+
+// Generate generates a new *Error with the same class and code, and new arguments.
+func (e *Error) Generate(args ...interface{}) error {
+	return e.levelGeneratef(1, e.message, args...)
+}
+
+// Generatef generates a new *Error with the same class and code, and a new formatted message.
+func (e *Error) Generatef(format string, args ...interface{}) error {
+	return e.levelGeneratef(1, format, args...)
+}
+
+// levelGeneratef is an inner interface to generate new *Error
+func (e *Error) levelGeneratef(stackSkipLevel int, format string, args ...interface{}) error {
```

Comment:
this function name will make some confused, because `Error` contains a `level`, but this function is mainly used to  set the stack level.